pkg/config/sync.go

```diff
@@ -0,0 +1,207 @@
+//
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	routeV1 "github.com/openshift/api/route/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	controller "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/infrastructure"
+)
+
+const (
+	OperatorConfigName     = "devworkspace-operator-config"
+	openShiftTestRouteName = "devworkspace-controller-test-route"
+)
+
+var (
+	Routing         *controller.RoutingConfig
+	Workspace       *controller.WorkspaceConfig
+	internalConfig  *controller.OperatorConfiguration
+	configMutex     sync.Mutex
+	configNamespace string
+)
+
+func SetConfigForTesting(config *controller.OperatorConfiguration) {
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	mergeConfig(config, internalConfig)
+	updatePublicConfig()
+}
+
+func SetupControllerConfig(client crclient.Client) error {
+	if internalConfig != nil {
+		return fmt.Errorf("internal controller configuration is already set up")
+	}
+	internalConfig = &controller.OperatorConfiguration{}
+	namespace, err := infrastructure.GetNamespace()
+	if err != nil {
+		return err
+	}
+	configNamespace = namespace
+	config, err := getClusterConfig(configNamespace, client)
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		internalConfig = DefaultConfig.DeepCopy()
+		updatePublicConfig()
+	} else {
+		syncConfigFrom(config)
+	}
+	if internalConfig.Routing.ClusterHostSuffix == "" && infrastructure.IsOpenShift() {
+		if err := fillOpenShiftRouteSuffix(client); err != nil {
+			return err
+		}
+		updatePublicConfig()
+	}
+	return nil
+}
+
+func ExperimentalFeaturesEnabled() bool {
+	if internalConfig.EnableExperimentalFeatures == nil {
+		return false
+	}
+	return *internalConfig.EnableExperimentalFeatures
+}
+
+func getClusterConfig(namespace string, client crclient.Client) (*controller.DevWorkspaceOperatorConfig, error) {
+	clusterConfig := &controller.DevWorkspaceOperatorConfig{}
+	if err := client.Get(context.Background(), types.NamespacedName{Name: OperatorConfigName, Namespace: namespace}, clusterConfig); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return clusterConfig, nil
+}
+
+func syncConfigFrom(newConfig *controller.DevWorkspaceOperatorConfig) {
+	if newConfig == nil || newConfig.Name != OperatorConfigName || newConfig.Namespace != configNamespace {
+		return
+	}
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	mergeConfig(newConfig.Config, internalConfig)
+	updatePublicConfig()
+}
+
+func restoreDefaultConfig() {
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	updatePublicConfig()
+}
+
+func updatePublicConfig() {
```

Comment:
Maybe it makes sense to report in logs when changes are detected and Config is reloaded? Probably also print the result how it's merged with default

Reply:
I actually had this for debugging and removed it before opening the PR :D 

I'll re-add

Reply:
Added. Currently, we log non-default values only (otherwise it can be very long). 

One question: should we be logging clusterHostSuffix? Someone using DWO might want to suppress the URL of their cluster in logs, especially if sharing publicly in e.g. a bug report.